File: pymath/matrix.py
def determinant(matrix):
    """Compute the determinant of a square matrix using recursion."""
    if len(matrix) == 1:
        return matrix[0][0]

    det = 0
    for col in range(len(matrix)):
        # Compute the minor matrix by removing the first row and current column
        minor = [row[:col] + row[col + 1:] for row in matrix[1:]]
        # Apply Laplace expansion
        det += (-1) ** col * matrix[0][col] * determinant(minor)

    return det


def hermitian_conjugate(matrix):
    """Compute the Hermitian conjugate (conjugate transpose) of a complex matrix."""
    return [[(x[0], -x[1]) for x in col] for col in zip(*matrix)]


def matrix_multiply(A, B):
    """Multiply two matrices A and B."""
    return [
        [sum(a * b for a, b in zip(A_row, B_col)) for B_col in zip(*B)]
        for A_row in A
    ]


def matrix_exponential(matrix, iterations=10):
    """Compute the matrix exponential using the power series expansion."""
    n = len(matrix)

    # Initialize the result as an identity matrix
    result = [[(1.0 if i == j else 0.0, 0.0) for j in range(n)] for i in range(n)]
    current_term = [[(x[0], x[1]) for x in row] for row in result]

    for k in range(1, iterations + 1):
        # Compute the next term in the power series
        current_term = matrix_multiply(current_term, matrix)
        current_term = [[(x[0] / k, x[1] / k) for x in row] for row in current_term]

        # Add the term to the result
        for i in range(n):
            for j in range(n):
                result[i][j] = (result[i][j][0] + current_term[i][j][0],
                                result[i][j][1] + current_term[i][j][1])

    return result


def matrix_power(matrix, power):
    """Raise a square matrix to an integer power using exponentiation by squaring."""
    if not matrix:
        return []

    n = len(matrix)

    # Ensure the matrix is square
    if any(len(row) != n for row in matrix):
        raise ValueError("Matrix must be square")

    # Initialize the identity matrix
    result = [[1.0 if i == j else 0.0 for j in range(n)] for i in range(n)]
    current = [row.copy() for row in matrix]
    exponent = abs(power)

    while exponent > 0:
        if exponent % 2 == 1:
            result = matrix_multiply(result, current)
        current = matrix_multiply(current, current)
        exponent //= 2

    # Handle negative exponent by computing the inverse
    if power < 0:
        result = matrix_inverse(result)

    return result


def matrix_inverse(matrix):
    """Compute the inverse of a square matrix using the Gauss-Jordan elimination method."""
    n = len(matrix)

    # Augment the matrix with the identity matrix
    augmented = [row + [1.0 if i == j else 0.0 for j in range(n)] for i, row in enumerate(matrix)]

    for i in range(n):
        # Normalize the pivot row
        pivot = augmented[i][i]
        if pivot == 0:
            raise ValueError("Matrix is singular and cannot be inverted")

        for j in range(2 * n):
            augmented[i][j] /= pivot

        # Eliminate other entries in the current column
        for k in range(n):
            if k != i and augmented[k][i] != 0:
                factor = augmented[k][i]
                for j in range(2 * n):
                    augmented[k][j] -= factor * augmented[i][j]

    # Extract the inverse matrix
    return [row[n:] for row in augmented]

File: src/matrix.cpp
#include "matrix.hpp"
#include <cmath>
#include <algorithm>

namespace matrix {

    void check_square(const Matrix& matrix) {
        if (matrix.empty() || matrix.size() != matrix[0].size()) {
            throw std::invalid_argument("Matrix must be square");
        }
    }

    Matrix multiply_scalar(const Matrix& matrix, const Complex& scalar) {
        Matrix result = matrix;
        for (auto& row : result) {
            for (auto& element : row) {
                element *= scalar;
            }
        }
        return result;
    }

    Matrix identity_matrix(size_t n) {
        Matrix I(n, std::vector<Complex>(n, 0.0));
        for (size_t i = 0; i < n; ++i) {
            I[i][i] = 1.0;
        }
        return I;
    }

    Matrix matrix_minor(const Matrix& matrix, size_t row, size_t col) {
        Matrix minor;
        for (size_t i = 0; i < matrix.size(); ++i) {
            if (i == row) continue;
            std::vector<Complex> new_row;
            for (size_t j = 0; j < matrix[i].size(); ++j) {
                if (j != col) new_row.push_back(matrix[i][j]);
            }
            minor.push_back(new_row);
        }
        return minor;
    }

    Complex determinant(const Matrix& matrix) {
        check_square(matrix);
        if (matrix.size() == 1) return matrix[0][0];
        
        Complex det = 0.0;
        for (size_t col = 0; col < matrix[0].size(); ++col) {
            Complex sign = (col % 2 == 0) ? 1.0 : -1.0;
            det += matrix[0][col] * sign * determinant(matrix_minor(matrix, 0, col));
        }
        return det;
    }

    Matrix hermitian_conjugate(const Matrix& matrix) {
        Matrix result(matrix[0].size(), std::vector<Complex>(matrix.size()));
        for (size_t i = 0; i < matrix.size(); ++i) {
            for (size_t j = 0; j < matrix[i].size(); ++j) {
                result[j][i] = std::conj(matrix[i][j]);
            }
        }
        return result;
    }

    Matrix multiply(const Matrix& A, const Matrix& B) {
        check_square(A);
        check_square(B);
        size_t n = A.size();
        Matrix result(n, std::vector<Complex>(n, 0.0));
        
        for (size_t i = 0; i < n; ++i) {
            for (size_t j = 0; j < n; ++j) {
                for (size_t k = 0; k < n; ++k) {
                    result[i][j] += A[i][k] * B[k][j];
                }
            }
        }
        return result;
    }

    Matrix exponential(const Matrix& matrix, int iterations) {
        check_square(matrix);
        size_t n = matrix.size();
        Matrix result = identity_matrix(n);
        Matrix current_term = result;
        
        for (int k = 1; k <= iterations; ++k) {
            current_term = multiply(current_term, matrix);
            for (size_t i = 0; i < n; ++i) {
                for (size_t j = 0; j < n; ++j) {
                    result[i][j] += current_term[i][j] / static_cast<double>(k);
                }
            }
        }
        return result;
    }

    Matrix power(const Matrix& matrix, int power) {
        check_square(matrix);
        if (power == 0) return identity_matrix(matrix.size());
        
        Matrix result = identity_matrix(matrix.size());
        Matrix base = matrix;
        int exponent = abs(power);
        
        while (exponent > 0) {
            if (exponent % 2 == 1) {
                result = multiply(result, base);
            }
            base = multiply(base, base);
            exponent /= 2;
        }
        return (power < 0) ? inverse(result) : result;
    }

    Matrix inverse(const Matrix& matrix) {
        check_square(matrix);
        size_t n = matrix.size();
        Complex det = determinant(matrix);
        if (std::abs(det) < 1e-10) {
            throw std::invalid_argument("Matrix is singular");
        }
        
        Matrix adjugate(n, std::vector<Complex>(n));
        for (size_t i = 0; i < n; ++i) {
            for (size_t j = 0; j < n; ++j) {
                Complex sign = ((i + j) % 2 == 0) ? 1.0 : -1.0;
                adjugate[j][i] = sign * determinant(matrix_minor(matrix, i, j));
            }
        }
        return multiply_scalar(adjugate, Complex(1.0)/det);
    }
}